### ISSUE TITLE
LCH tool handles saturated yellows very poorly

### ISSUE THREAD

**Issue by chris-morgan** [LeaVerou/css.land#9]
I have no idea exactly what “right” behaviour here is, but the current behaviour doesn’t *look* right, and certainly *feels* terrible.

In short: yellows are handled *really* badly in this tool https://css.land/lch/.

As an example: import `#ff0`. This comes through as lch(97.607% 94.708 99.575), rgb(100% 100% 0%) and color(display-p3 1 0.9829 0.7277) (“actually color(display-p3 1.0001 1 0.3309)”).

The Display P3 value bounding there seem strange, with the blue channel 0.3309 → 0.7277, but I’m not qualified to judge it truly. I just remark on the apparent strangeness of it.

(The three bars are *way* off base, showing colours much lighter than the yellow actually is. This yellow is nowhere *near* perceptually equivalent to other colours at 97.607% lightness.)

Try twiddling the chroma channel. Take it *any* higher or lower, and either the red or the green channel will *slightly* exceed 100%, and it blitzes the yellow, lightening it 75% (to the vicinity of #ffffc0).

How about an only somewhat lighter yellow? Import `#ff6`. You get lch(97.931% 71.801 100.853), rgb(100% 100% 40%), color(display-p3 1 0.9896 0.7048) (“actually color(display-p3 1.0001 1 0.4961)”).

Again, strange Display P3 blue, but meh.

Try twiddling the chroma channel again. The same problem occurs.

For these #ff_ values, this is all happening because RGB isn’t straight relative to LCH: lightening the yellow in RGB space is causing the LCH hue to rotate a little, and so red or green slightly exceeds 100%.

But even for slightly less extreme yellows picked manually, as soon as they go out of range, the technique used to restore them into the sRGB range destroys the colour utterly.

Vibrant yellow is such a weird colour and hard to deal with. For that matter, vibrant colours in general are surprisingly hard to deal with in perceptually-uniform colour spaces, because you’re right on the boundaries of what’s possible. That’s something they don’t tend to tell you.

I don’t know what’s at fault here, whether it’s this tool only or whether browser implementations of `lch()` would be bad as well.

**Comment by LeaVerou** [comment]
Indeed, gamut mapping by chroma reduction is terrible for yellows. This is established in the literature, @svgeesus was reading about it in *Chroma Gamut Mapping, Jan Morovic, Wiley 2008* a couple days ago, perhaps he can cite a more specific page. We are currently looking for a better algorithm.

**Comment by svgeesus** [comment]
The issue here is that the tool tries to bring a color in-gamut in LCH space by holding L and H constant and reducing C until the resulting color is inside the destination gamut. Normally this works fairly well.

Looking at the P3 gamut [projected onto the (a,b or (C, H) plane](https://drafts.csswg.org/css-color-4/#valdef-color-display-p3) we see that P3 yellow is very similar in hue to sRGB yellow and both have a similar lightness of 97% but P3 yellow has significantly higher chroma (123 vs. 94). After some chroma reduction the color is _almost_ in gamut but the red is still slightly greater than one. The surface of the sRGB gamut volume is very shallow at that point, and the constant-chroma ray is skating above the surface. And so the chroma is reduced and reduced (and thus, all three channels tend towards equal values, which would produce a grey).

What we are considering is, at each chroma reduction step, checking to see if we are close to the gamut boundary by calculating the deltaE between the current color and that same color with per-component clipping. If the deltaE is such that the per-component-clipped one is perceptually close, we stop there and clip.

Another option, which tends to be used when the destination black point is higher (or the destination white point lower) than the source gamut, is to tilt the constant-chroma ray a little up (for very dark colors) or a little down (for very light ones) precisely to avoid this skating over the top (or bottom) of the gamut volume.

Lastly, there are some gamut mapping methods which use hand-chosen rules for different areas of the gamut. Normally this is prohibitively expensive but it might be worth experimenting for the very common cases (and `display-p3` to `srgb` is certainly a common case).

> Chroma Gamut Mapping, Jan Morovic, Wiley 2008
Color Gamut Mapping, Jan Morovic, Wiley 2008

**Comment by svgeesus** [comment]
I did some calculations. deltaE2000 between P3 yellow and sRGB yellow is 5 (noticeable side by side, but fairly similar) while deltaE2000 between P3 yellow and our super-desaturated gamut mapped yellow is a whopping 22.4.

![yellows](https://user-images.githubusercontent.com/2506926/83930217-75043680-a764-11ea-9175-98dfd6ae52c2.png)
